python/ray/_private/pydantic_compat.py

```diff
@@ -0,0 +1,107 @@
+# flake8: noqa
+import weakref
+from pkg_resources import packaging
+from typing import Any
+
+# Pydantic is a dependency of `ray["default"]` but not the minimal installation,
+# so handle the case where it isn't installed.
+try:
+    import pydantic
+
+    PYDANTIC_INSTALLED = True
+except ImportError:
+    pydantic = None
+    PYDANTIC_INSTALLED = False
+
+
+class CloudpickleableSchemaSerializer:
+    def __init__(self, schema, core_config):
+        self._schema = schema
+        self._core_config = core_config
+
+        from pydantic_core import SchemaSerializer
+
+        self._schema_serializer = SchemaSerializer(schema, core_config)
+
+    def __reduce__(self):
+        return CloudpickleableSchemaSerializer, (self._schema, self._core_config)
+
+    def __getattr__(self, attr: str):
+        return getattr(self._schema_serializer, attr)
+
+
+class WeakRefWrapper:
+    def __init__(self, obj: Any):
+        if obj is None:
+            self._wr = None
+        else:
+            self._wr = weakref.ref(obj)
+
+    def __reduce__(self):
+        return WeakRefWrapper, (self(),)
+
+    def __call__(self) -> Any:
+        if self._wr is None:
+            return None
+        else:
+            return self._wr()
+
+
+def monkeypatch_pydantic_2_for_cloudpickle():
```

Comment:
I guess this monkey patch makes pydantic 2.x work but the patch in `python/ray/util/serialization_addons.py:register_pydantic_serializer` is only meant for pydantic 1.x.

Should we combine these into a single file (move the pydantic 1.x patch into this file)?

Reply:
Eh, for now I'm trying to minimize the size of the diff (there will be more changes to come to fully enable compatibility). So I'd prefer not to touch that part for now.